src/main/angular/app/services/graph-processor.service.spec.ts

```diff
@@ -0,0 +1,236 @@
+        const allProjects: Project[] = await fakeDataService.listProjects();
+        // Add the projects
+        changes.projects = new SimpleChange([], allProjects, true);
+        await service.processChanges(changes, properties, fakeDataService);
+
+        // Remove the first project
+        const projects: Project[] = allProjects
+          .filter((value, index) => index !== 0)
+          .map(value => value);
+        const projectData: ProjectGraphData[] = await Promise.all(
+          projects.map(project =>
+            fakeDataService.getProjectGraphData(project.projectId)
+          )
+        );
+        changes.projects = new SimpleChange(allProjects, projects, false);
+        await service.processChanges(changes, properties, fakeDataService);
+
+        const expectedColumns = 1 + projects.length * 3;
+        expect(properties.columns.length).toBe(expectedColumns);
+        expect(properties.columns).not.toContain(allProjects[0].projectId);
+        properties.graphData.forEach(row => {
+          if (row[0] instanceof Date) {
+            const time = row[0].getTime();
+            expect(row.length).toBe(expectedColumns);
+            projectData.forEach((data, index) => {
+              expect(row[1 + 3 * index]).toBe(
+                data.dateToNumberIAMBindings[time]
+              );
+            });
+          } else {
+            fail(`Row ${row} doesn't have a time`);
+          }
+        });
+      });
+      it('Removes multiple projects successfully', async () => {
+        const allProjects: Project[] = await fakeDataService.listProjects();
+        // Add the projects
+        changes.projects = new SimpleChange([], allProjects, true);
+        await service.processChanges(changes, properties, fakeDataService);
+
+        // Remove all but the first two projects
+        const projects: Project[] = allProjects
+          .filter((value, index) => [0, 1].includes(index))
+          .map(value => value);
+        const projectData: ProjectGraphData[] = await Promise.all(
+          projects.map(project =>
+            fakeDataService.getProjectGraphData(project.projectId)
+          )
+        );
+        changes.projects = new SimpleChange(allProjects, projects, false);
+        await service.processChanges(changes, properties, fakeDataService);
+
+        const expectedColumns = 1 + projects.length * 3;
+        expect(properties.columns.length).toBe(expectedColumns);
+        for (let i = 2; i < allProjects.length; i++) {
+          expect(properties.columns).not.toContain(allProjects[i].projectId);
+        }
+
+        properties.graphData.forEach(row => {
+          if (row[0] instanceof Date) {
+            const time = row[0].getTime();
+            expect(row.length).toBe(expectedColumns);
+            projectData.forEach((data, index) => {
+              expect(row[1 + 3 * index]).toBe(
+                data.dateToNumberIAMBindings[time]
+              );
+            });
+          } else {
```

Comment:
General concepts to consider: 
1. Structure testing into 3 blocks: Arrange, Act, Assert. 
2. Avoid programmatic logic in tests (for each assert, etc). Even though it's more code, the intent is always very clear.

Reply:
OK I all but rewrote the tests. How is the format they're in now?